File: backend/app/schemas/commerce.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from app.schemas.reviews import ReviewIntelligenceDTO

class ProductSpecDTO(BaseModel):
    key: str
    value: str

class ProductDTO(BaseModel):
    id: str
    name: str
    brand: str
    category: str
    price: float
    original_price: Optional[float] = None
    currency: str = "INR"
    rating: float = 4.8
    reviews_count: int = 124
    image_url: str
    tagline: str
    description: str
    features: List[str] = []
    specs: List[ProductSpecDTO] = []
    pros: List[str] = []
    cons: List[str] = []
    stock_status: Optional[str] = "In Stock (Available)"
    delivery_eta: Optional[str] = "Tomorrow by 5:00 PM via Delhivery"
    merchant_trust_score: Optional[float] = 98.5
    in_stock: bool = True
    delivery_time: str = "2-3 business days"
    gst_rate_pct: float = 18.0
    price_tiers: Optional[List[Dict[str, Any]]] = None
    price_tiers_json: Optional[str] = None
    review_sentiment_score: Optional[float] = 0.90
    popularity_score: Optional[float] = 0.88
    match_score: Optional[float] = None
    ranking_breakdown: Optional[Dict[str, float]] = None
    why_recommended: Optional[str] = None
    review_intelligence: Optional[ReviewIntelligenceDTO] = None


class CartItemDTO(BaseModel):
    product_id: str
    name: str
    price: float
    quantity: int = 1
    image_url: str
    category: Optional[str] = None
    tier_used: Optional[Dict[str, Any]] = None
    discount_amount: float = 0.0
    effective_price: Optional[float] = None

class CartDTO(BaseModel):
    items: List[CartItemDTO] = []
    subtotal: float = 0.0
    tax_gst: float = 0.0
    shipping: float = 0.0
    discount: float = 0.0
    coupon_applied: Optional[str] = None
    total: float = 0.0
    currency: str = "INR"

class ComparisonAttributeDTO(BaseModel):
    attribute: str
    values: Dict[str, str]

class ComparisonDataDTO(BaseModel):
    product_ids: List[str]
    products: List[ProductDTO]
    attributes: List[ComparisonAttributeDTO]
    verdict: str

class ChatMessageDTO(BaseModel):
    role: str # "user" | "assistant"
    content: str
    timestamp: Optional[str] = None
    recommended_products: Optional[List[Any]] = None
    comparison_data: Optional[ComparisonDataDTO] = None
    suggested_prompts: Optional[List[str]] = None
    action_type: Optional[str] = None # "add_to_cart" | "view_cart" | "compare" | "checkout" | "select_product" | "select_address"
    flow_step: Optional[str] = None
    ai_recommendation_reason: Optional[Dict[str, Any]] = None
    selected_product: Optional[Any] = None
    selected_address: Optional[Dict[str, Any]] = None
    order_summary: Optional[Dict[str, Any]] = None
    saved_addresses: Optional[List[Dict[str, Any]]] = None
    review_intelligence: Optional[ReviewIntelligenceDTO] = None
    before_checkout_summary: Optional[str] = None

class CommerceChatRequestDTO(BaseModel):
    query: str
    history: List[ChatMessageDTO] = []
    cart: Optional[CartDTO] = None
    action: Optional[str] = None # "select_product" | "select_address" | "confirm_autopay_purchase"
    selected_product_id: Optional[str] = None
    selected_address: Optional[Dict[str, Any]] = None
    quantity: Optional[int] = 1

class CommerceChatResponseDTO(BaseModel):
    message: str
    recommended_products: List[ProductDTO] = []
    comparison_data: Optional[ComparisonDataDTO] = None
    suggested_prompts: List[str] = []
    cart: Optional[CartDTO] = None
    action_triggered: Optional[str] = None
    flow_step: Optional[str] = None # "TOP_RECOMMENDATIONS" | "ADDRESS_SELECTION" | "ORDER_SUMMARY" | "AUTONOMOUS_PURCHASE" | "APPROVAL_REQUIRED"
    ai_recommendation_reason: Optional[Dict[str, Any]] = None
    recommendation_reason: Optional[str] = None
    confidence_score: Optional[float] = None
    selected_product: Optional[ProductDTO] = None
    selected_address: Optional[Dict[str, Any]] = None
    order_summary: Optional[Dict[str, Any]] = None
    saved_addresses: Optional[List[Dict[str, Any]]] = None
    checkout_link: Optional[str] = None
    autonomous_order: Optional[Dict[str, Any]] = None
    requires_approval: Optional[bool] = False
    autopay_guardrail_info: Optional[Dict[str, Any]] = None
    parsed_intent: Optional[Dict[str, Any]] = None
    review_intelligence: Optional[ReviewIntelligenceDTO] = None
    before_checkout_summary: Optional[str] = None


class AdvisorParsedIntentDTO(BaseModel):
    intent: str = "product_recommendation"
    category: Optional[str] = None
    budget: Optional[float] = None
    desired_specs: List[str] = []
    rating_min: Optional[float] = None
    preferred_brands: List[str] = []
    raw_query: Optional[str] = None

class AdvisorRecommendRequestDTO(BaseModel):
    query: str
    budget_cap: Optional[float] = None
    limit: Optional[int] = 3

class AdvisorRecommendationResponseDTO(BaseModel):
    recommended_products: List[ProductDTO] = []
    recommendation_reason: str = ""
    confidence_score: float = 0.0
    parsed_intent: Optional[AdvisorParsedIntentDTO] = None
    query: Optional[str] = None

class CheckoutRequestDTO(BaseModel):
    cart: CartDTO
    customer_email: Optional[str] = "finance.ops@acmedirect.com"
    customer_name: Optional[str] = "Acme Direct Corp"
    customer_phone: Optional[str] = "+91 98765 43210"

class CheckoutResponseDTO(BaseModel):
    payment_link_id: str
    payment_url: str
    order_id: str
    amount: float
    currency: str = "INR"
    status: str = "created"
    qr_code_mock: str
    expires_at: str
    summary_items: List[CartItemDTO] = []

File: apps/customer-storefront/src/types/commerce.ts
export interface ProductSpec {
  key: string;
  value: string;
}

export interface Product {
  id: string;
  sku?: string;
  name: string;
  brand?: string;
  category: string;
  price: number; // Customer-facing GST-inclusive price
  customer_price?: number;
  base_price?: number; // Net price excluding GST
  gst_rate?: number; // e.g. 0.18
  gst_rate_pct?: number; // e.g. 18.0
  gst_amount?: number;
  price_display?: string;
  cost_price?: number;
  original_price?: number;
  currency?: string;
  rating?: number;
  reviews_count?: number;
  image_url: string;
  tagline?: string;
  description: string;
  features?: string[];
  key_features?: string[];
  specs?: ProductSpec[];
  pros?: string[];
  cons?: string[];
  stock?: number;
  stock_quantity?: number;
  stock_status?: string;
  delivery_eta?: string;
  merchant_trust_score?: number;
  inventory_status?: 'IN_STOCK' | 'LOW_STOCK' | 'OUT_OF_STOCK' | 'PRE_ORDER' | 'DISCONTINUED';
  in_stock?: boolean;
  offer?: string;
  active_offer?: string;
  delivery_time?: string;
}

export interface ProductCategoryStats {
  name: string;
  count: number;
}

export interface ProductStats {
  total_products: number;
  total_valuation_inr?: number;
  total_inventory_value?: number;
  in_stock_rate_pct?: number;
  in_stock_count?: number;
  out_of_stock_count?: number;
  active_offers_count: number;
  total_categories?: number;
  categories?: ProductCategoryStats[];
}

export interface ProductListResponse {
  items: Product[];
  total: number;
  limit: number;
  offset: number;
}

export interface CartItem {
  product_id: string;
  name: string;
  price: number;
  quantity: number;
  image_url: string;
  category?: string;
  product?: Product;
}

export interface CartState {
  items: CartItem[];
  items_total?: number;
  subtotal: number;
  delivery_fee?: number;
  platform_fee?: number;
  gst_included?: number;
  tax_gst: number;
  shipping: number;
  discount: number;
  coupon_applied?: string | null;
  total: number;
  currency: string;
}

export interface ComparisonAttribute {
  attribute: string;
  values: Record<string, string>;
}

export interface ComparisonData {
  product_ids?: string[];
  products: Product[];
  attributes?: ComparisonAttribute[];
  features?: any[];
  title?: string;
  verdict: string;
}

export interface DeliveryAddress {
  id: string;
  label: string;
  recipient_name: string;
  address_line: string;
  city: string;
  state: string;
  pincode: string;
  phone: string;
  is_default?: boolean;
}

export interface AIRecommendationReason {
  recommended_product_id: string;
  product_name: string;
  headline: string;
  why_bullets: string[];
}

export interface AdvisorOrderSummary {
  product_id: string;
  product_name: string;
  product_image: string;
  brand: string;
  quantity: number;
  unit_price: number;
  base_subtotal: number;
  gst_amount: number;
  delivery_fee: number;
  total_amount: number;
  delivery_address: DeliveryAddress;
  expected_delivery: string;
  within_limit: boolean;
  within_budget: boolean;
  payment_method: string;
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: string;
  action_type?: string;
  checkout_link?: string;
  flow_step?: string;
  recommended_products?: Product[];
  comparison_data?: ComparisonData | null;
  ai_recommendation_reason?: AIRecommendationReason | null;
  selected_product?: Product | null;
  selected_address?: DeliveryAddress | null;
  order_summary?: AdvisorOrderSummary | null;
  saved_addresses?: DeliveryAddress[] | null;
  suggested_prompts?: string[];
  autonomous_order?: any;
  requires_approval?: boolean;
}

export interface CommerceChatResponse {
  message: string;
  action_triggered?: string;
  action_type?: string;
  flow_step?: string;
  checkout_link?: string;
  cart?: any;
  recommended_products?: Product[];
  comparison?: ComparisonData | null;
  comparison_data?: ComparisonData | null;
  ai_recommendation_reason?: AIRecommendationReason | null;
  selected_product?: Product | null;
  selected_address?: DeliveryAddress | null;
  order_summary?: AdvisorOrderSummary | null;
  saved_addresses?: DeliveryAddress[] | null;
  suggested_prompts?: string[];
  autonomous_order?: any;
  requires_approval?: boolean;
  autopay_guardrail_info?: any;
}

export interface CheckoutResult {
  order_id: string;
  payment_link?: string;
  payment_link_id?: string;
  payment_url?: string;
  qr_code_data?: string;
  qr_code_mock?: string;
  amount: number;
  currency: string;
  status: string;
  receipt?: string;
}
